### PR TITLE
Added FrozenList subtraction

### LINKED ISSUE #15475: Groupby all levels except the specified ones
Inspired by 

> http://stackoverflow.com/questions/25606478/grouping-by-everything-except-for-one-index-column-in-pandas

 I was thinking if it would be useful to add a parameter to the groupby function (e.g. allbut=True) so that pandas will group all levels except the specified ones) .

I suggest so because in a multilevel index with several levels, eliminating just one level requires a quite verbose approach

### PR DESCRIPTION
**Pull request by benthayer**
- [x] closes #15475
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/indexes/frozen.py [hunk 1]
```diff
@@ -31,6 +31,11 @@ def __add__(self, other):
 
     __iadd__ = __add__
 
+    def __sub__(self, other):
+        other = set(other)
+        temp = [x for x in self if x not in other]
+        return self.__class__(temp)
```
Comment (jreback) [inline pandas/indexes/frozen.py]: add isub as well (and a test)

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -26,6 +26,8 @@ Check the :ref:`API Changes <whatsnew_0200.api_breaking>` and :ref:`deprecations
 New features
 ~~~~~~~~~~~~
 
+- Added subtraction from FrozenLists (:issue:`15475`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: this is called difference 

i think might be nice to add an example to th docs as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/groupby.rst [hunk 3]
```diff
@@ -126,6 +126,14 @@ We could naturally group by either the ``A`` or ``B`` columns or both:
    grouped = df.groupby('A')
    grouped = df.groupby(['A', 'B'])
 
+If we also have a MultiIndex on columns ``A`` and ``B``, we can group by all
+but the specified columns
```
Comment (jreback) [inline doc/source/groupby.rst]: add a versionadded tag here

File: pandas/tests/indexes/test_frozen.py [hunk 4]
```diff
@@ -23,13 +23,30 @@ def test_add(self):
...
+        result = self.container - [2]
+        expected = FrozenList([1, 3, 4, 5])
+        self.check_result(result, expected)
+
+    def test_sub_dupe(self):
+        result = FrozenList([1, 2, 3, 2]) - [2]
+        expected = FrozenList([1, 3])
+        self.check_result(result, expected)
+
     def test_inplace(self):
```
Comment (jreback) [inline pandas/tests/indexes/test_frozen.py]: can you rename this to make consistent

**Comment by jorisvandenbossche** [comment]
I want to raise some concern regarding adding this feature (without wanting to dismiss your effort @bthayer2365 !). I think we should think about this some more. 

Adding this would introduce a discrepancy between index level names and column names, while we actually try to lessen the gap (eg by being able to specify an index level name in places where you can specify a column name). This is actually a feature we *removed* from column names, in favor of the `difference` method. 
So that is maybe also an alternative? Putting this in a method instead of overloading subtraction? (although I certainly agree what we try to obtain here is useful, and as a method it is more verbose)

**Comment by jreback** [comment]
no conceptual problem with making these real setops, e.g. ``.union``, ``.difference`` as @jorisvandenbossche suggests. Publicly these are *only* used for names of levels anyhow.

**Comment by bthayer2365** [comment]
Would you like me to change `__add__` to `union` as well?

**Comment by jreback** [comment]
yes; u will have to deprecate the add though

**Comment by bthayer2365** [comment]
To do that, do I just decorate it with `@deprecated`, or is there more?

**Comment by jreback** [comment]
that should work (and add in whatsnew as well)
iadd too

File: doc/source/groupby.rst [hunk 5]
```diff
@@ -126,6 +126,16 @@ We could naturally group by either the ``A`` or ``B`` columns or both:
...
    grouped = df.groupby(['A', 'B'])
 
+.. versionadded:: 0.20
+
+If we also have a MultiIndex on columns ``A`` and ``B``, we can group by all
+but the specified columns
+
+.. ipython:: python
+
+   df.set_index(['A', 'B'])
```
Comment (jreback) [inline doc/source/groupby.rst]: this won't work as it returns a new object, instead something like

```
df2 = df.set_index(['A', 'B'])
grouped = df2.groupby(level=....)
```

File: doc/source/whatsnew/v0.20.0.txt [hunk 6]
```diff
@@ -26,6 +26,8 @@ Check the :ref:`API Changes <whatsnew_0200.api_breaking>` and :ref:`deprecations
 New features
 ~~~~~~~~~~~~
 
+- Added difference from FrozenLists (:issue:`15475`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: ``.difference()`` method for ``FrozenList``

File: pandas/indexes/frozen.py [hunk 7]
```diff
@@ -25,6 +27,7 @@ class FrozenList(PandasObject, list):
     #           typechecks
 
     def __add__(self, other):
+        warnings.warn("__add__ is deprecated, use union(...)", FutureWarning)
         if isinstance(other, tuple):
```
Comment (jreback) [inline pandas/indexes/frozen.py]:
```
return self.union(other)
```

File: pandas/indexes/frozen.py [hunk 8]
```diff
@@ -80,6 +83,16 @@ def __repr__(self):
     __setitem__ = __setslice__ = __delitem__ = __delslice__ = _disabled
     pop = append = extend = remove = sort = insert = _disabled
 
+    def union(self, other):
+        if isinstance(other, tuple):
```
Comment (jreback) [inline pandas/indexes/frozen.py]: can you add a doc-string

File: pandas/indexes/frozen.py [hunk 9]
```diff
@@ -80,6 +83,16 @@ def __repr__(self):
...
+            other = list(other)
+        return self.__class__(super(FrozenList, self).__add__(other))
+
+    def difference(self, other):
+        other = set(other)
```
Comment (jreback) [inline pandas/indexes/frozen.py]: same here

File: pandas/indexes/frozen.py [hunk 10]
```diff
@@ -80,6 +83,16 @@ def __repr__(self):
     __setitem__ = __setslice__ = __delitem__ = __delslice__ = _disabled
     pop = append = extend = remove = sort = insert = _disabled
 
+    def union(self, other):
+        if isinstance(other, tuple):
+            other = list(other)
+        return self.__class__(super(FrozenList, self).__add__(other))
```
Comment (jreback) [inline pandas/indexes/frozen.py]: I *think* ``__iadd__`` is also defined? if so, deprecate that as well.

**Comment by bthayer2365** [comment]
I tried resolving the conflicts and merging manually, but the check failed, so I reverted back to before I resolved the conflict. Am I supposed to resolve the merge conflicts and merge master?

**Comment by cpcloud** [comment]
@bthayer2365 Yep, you need to resolve conflicts on your local branch. You should do:

```
git pull upstream master --rebase --tags
```
And resolve the conflicts one by one, doing `git rebase --continue` once each one has been resolved. Then push to your fork.

**Comment by jreback** [comment]
yes the idea is to rebase on master. But don't worry about that as much as passing tests (of course I think travis / amazon is having issues atm).

**Comment by bthayer2365** [comment]
Okay, should we just wait a bit to see if the tests work when everything
starts working again? Everything worked on the command line

**Comment by bthayer2365** [comment]
I'll leave it like this and see if it passes

**Comment by jreback** [comment]
you'll want to rebase on master as travis was having some issues

File: doc/source/whatsnew/v0.20.0.txt [hunk 11]
```diff
@@ -534,6 +536,7 @@ Deprecations
 - ``Series.sortlevel`` and ``DataFrame.sortlevel`` have been deprecated in favor of ``Series.sort_index`` and ``DataFrame.sort_index`` (:issue:`15099`)
 - importing ``concat`` from ``pandas.tools.merge`` has been deprecated in favor of imports from the ``pandas`` namespace. This should only affect explict imports (:issue:`15358`)
 - ``Series/DataFrame/Panel.consolidate()`` been deprecated as a public method. (:issue:`15483`)
+- ``FrozenList.__add__`` has been depricated in favor of ``FrozenList.union`` (:issue: `15475`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: deprecated. Add for ``__iadd__`` here as well

File: pandas/indexes/frozen.py [hunk 12]
```diff
@@ -25,11 +27,12 @@ class FrozenList(PandasObject, list):
...
-        if isinstance(other, tuple):
-            other = list(other)
-        return self.__class__(super(FrozenList, self).__add__(other))
+        warnings.warn("__add__ is deprecated, use union(...)", FutureWarning)
+        return self.union(other)
 
-    __iadd__ = __add__
+    def __iadd__(self, other):
+        warnings.warn("__iadd__ is deprecated, use union(...)", FutureWarning)
+        return self.union(other)
```
Comment (jreback) [inline pandas/indexes/frozen.py]: this is technically wrong; its an in-place modify, and doesn't return anything. Can you add a test for validate this?

*maybe*

```
def __iadd__(self, other):
    warnings(.....)
     if isinstance(other, tuple):
            other = list(other)
    super(FrozenList, self).__iadd__(other)
```

might work

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by bthayer2365** [comment]
Anything else? All tests passed 👍

**Comment by jreback** [comment]
thanks @bthayer2365

**Comment by cpcloud** [comment]
@bthayer2365 Great job! Thanks for your contribution.

**Comment by bthayer2365** [comment]
Glad I could help!

**Pull request closed without merging**